crates/sui-indexer-alt-jsonrpc/src/lib.rs

```diff
@@ -2,274 +2,555 @@
+    pub fn new(
+        rpc_args: RpcArgs,
+        registry: &Registry,
+        cancel: CancellationToken,
+    ) -> anyhow::Result<Self> {
         let RpcArgs {
-            listen_address,
+            rpc_listen_address,
             max_rpc_connections,
         } = rpc_args;
 
+        let metrics = RpcMetrics::new(registry);
+
         let server = ServerBuilder::new()
             .http_only()
-            .max_connections(max_rpc_connections)
-            .build(listen_address)
-            .await
-            .context("Failed to bind server")?;
+            .max_connections(max_rpc_connections);
+
+        let schema = Project::new(
+            env!("CARGO_PKG_VERSION"),
+            "Sui JSON-RPC",
+            "A JSON-RPC API for interacting with the Sui blockchain.",
+            "Mysten Labs",
+            "https://mystenlabs.com",
+            "build@mystenlabs.com",
+            "Apache-2.0",
+            "https://raw.githubusercontent.com/MystenLabs/sui/main/LICENSE",
+        );
 
         Ok(Self {
+            rpc_listen_address,
             server,
-            modules: RpcModule::new(()),
+            metrics,
+            modules: jsonrpsee::RpcModule::new(()),
+            schema,
+            cancel,
         })
     }
 
+    /// Return a copy of the metrics.
+    pub fn metrics(&self) -> Arc<RpcMetrics> {
+        self.metrics.clone()
+    }
+
     /// Add an [RpcModule] to the service. The module's methods are combined with the existing
     /// methods registered on the service, and the operation will fail if there is any overlap.
-    pub fn add_module<M>(&mut self, module: RpcModule<M>) -> anyhow::Result<()> {
+    pub fn add_module(&mut self, module: impl RpcModule) -> anyhow::Result<()> {
+        self.schema.add_module(module.schema());
         self.modules
-            .merge(module.remove_context())
+            .merge(module.into_impl().remove_context())
             .context("Failed to add module because of a name conflict")
     }
 
     /// Start the service (it will accept connections) and return a handle that will resolve when
     /// the service stops.
     pub async fn run(self) -> anyhow::Result<JoinHandle<()>> {
-        let Self { server, modules } = self;
+        let Self {
+            rpc_listen_address,
+            server,
+            metrics,
+            mut modules,
+            schema,
+            cancel,
+        } = self;
+
+        info!("Starting JSON-RPC service on {rpc_listen_address}",);
+        info!("Serving schema: {}", serde_json::to_string_pretty(&schema)?);
 
-        let listen_address = server
-            .local_addr()
-            .context("Can't start RPC service without listen address")?;
+        // Add a method to serve the schema to clients.
+        modules
+            .register_method("rpc.discover", move |_, _, _| json!(schema.clone()))
+            .context("Failed to add schema discovery method")?;
```

Comment:
@emmazzz, here we are adding the `rpc.discover` method that uses the results from calling `RpcModule::schema`.